### PR TITLE
[mini] Emit widen ops when storing to locals of size < 4. Fixes #58379.

### PR DESCRIPTION
**Pull request by kumpera**

### CODE REVIEW

**Comment by lewurm** [comment]
build interpreter

File: mono/mini/method-to-ir.c [hunk 1]
```diff
@@ -6607,10 +6607,78 @@ set_exception_type_from_invalid_il (MonoCompile *cfg, MonoMethod *method, unsign
...
+		if (cfg->cbb->last_ins == sp [0] && sp [0]->opcode == coerce_op) {
+			if (cfg->verbose_level > 2)
+				printf ("Found existing coercing that is enough\n");
+		} else {
+			MONO_INST_NEW (cfg, ins, coerce_op);
+			ins->dreg = alloc_ireg (cfg);
+			ins->sreg1 = sp [0]->dreg;
+			ins->type = STACK_I4;
+			ins->klass = mono_class_from_mono_type (header->locals [n]);
+	        MONO_ADD_INS (cfg->cbb, ins);
```
Comment (lewurm) [inline mono/mini/method-to-ir.c]: indent

File: mono/mini/iltests.il [hunk 2]
```diff
@@ -2896,4 +2896,58 @@ END:
...
+		ldloc.0
+		brfalse L_0
+
+		ldc.i4.1
+		ret
+L_0:
+		ldsfld int32 Tests::byte_val
+		stloc.1
+		ldloc.1
+		conv.i4
```
Comment (lewurm) [inline mono/mini/iltests.il]: why is `conv.i4` needed here?

File: mono/mini/iltests.il [hunk 3]
```diff
@@ -2896,4 +2896,58 @@ END:
...
+		.maxstack 1
+		.locals init (
+			unsigned int8	V_0,
+			int8	V_1,
+			int16	V_2,
+			unsigned int16	V_3)
+
+		ldc.i4 256
+		stsfld int32 Tests::byte_val
+		ldc.i4 65536
```
Comment (lewurm) [inline mono/mini/iltests.il]: looks like an arbitrary number, use hex literals 😛

**Comment by lewurm** [comment]
interpreter did it right :partyparrot:

File: mono/mini/iltests.il (hunk 3 above)
Comment (kumpera) [inline mono/mini/iltests.il]: fixed, used hex literals and explained what the tests checks and how.

File: mono/mini/iltests.il (hunk 2 above)
Comment (kumpera) [inline mono/mini/iltests.il]: no good reason, removed.

File: mono/mini/method-to-ir.c (hunk 1 above)
Comment (kumpera) [inline mono/mini/method-to-ir.c]: fixed

**Comment by lewurm** [comment]
build interpreter

**Comment by lewurm** [comment]
build interpreter

**Pull request merged**